sinks/src/main/scala/com.snowplowanalytics.snowplow.eventgen/Http.scala

```diff
@@ -0,0 +1,141 @@
+        headers
+          .map(kv => (kv._1, kv._2) match {
+            case (k, v) if k == "Origin" =>  
+              v.split(",")
+                .map(v => Raw(CIString(k), v))
+                .toSeq
+              
+              // Raw(CIString(k), v)
+            case (_, _) => Seq(Raw(CIString(kv._1), kv._2))
+          })
+          .toSeq
+          .flatten
+      }
+
+       // since /i requests have empty version, we pattern match to add the path.
+       // address is a var since we modify it when adding querystring
+       var address = generatedRequest.method.path.version match {
+        case "" => baseUri / generatedRequest.method.path.vendor
+        case version: String  => baseUri / generatedRequest.method.path.vendor / version
+       }
+
+      val body = generatedRequest.body match {
+        case Some(b) => b.toString()
+        case _       => ""
+      }
+
+      val req = generatedRequest.method match {
+
+        // POST requests: attach body, use the uri we already have, without adding querysring
+        case TrackerMethod.Post(_) =>
+          Request[F](method = Method.POST, uri = address).withEntity(body).putHeaders(parseHeaders(generatedRequest.headers))
+
+        // GET requests: add querystring, ignore body field
+        case TrackerMethod.Get(_) =>
+                    
+          // iterate querystring and add as kv pairs to the address
+          generatedRequest.qs match {
+            case None => address
+            case Some(querystring: HttpRequestQuerystring) => 
+              // val newUri = address
+              querystring.toProto
+              querystring.toProto.foreach(kv =>
+                address = address.withQueryParam(kv.getName(), kv.getValue())
+              )
+          }
+
+          Request[F](method = Method.GET, uri = address).putHeaders(parseHeaders(generatedRequest.headers))
+
+        // HEAD requests induce server errors that will take some digging to figure out.
+        // In the interest of getting to a usable tool quickly, for now we just throw an error for these.
+        case TrackerMethod.Head(_) => throw new NotImplementedError ( "HEAD requests not implemented" )
+
+        /*
+            generatedRequest.qs match {
+            case None => address
+            case Some(querystring: HttpRequestQuerystring) => 
+              // val newUri = address
+              querystring.toProto
+              querystring.toProto.foreach(kv =>
+                address = address.withQueryParam(kv.getName(), kv.getValue())
+              )
+          }
+          
+
+          Request[F](method = Method.HEAD, uri = address).putHeaders(parseHeaders(generatedRequest.headers))
+          */
+
+          // This hits org.http4s.client.UnexpectedStatus: unexpected HTTP status: 422 Unprocessable Entity for request HEAD 
+
+      }
+      // TODO: Some code repetition can probably be removed from this
+
+      return req
```

Comment:
Again, not a reason for changing anything. But maybe that's a personal preference, but I've always found a function should have a single reason to change. Never been a massive fan of internal methods especially as it makes them hard to test individually.

I'd structure this function in a way that address, body, parseHeaders are separate functions, this way buildRequest could potentially look like this:

```
def buildRequesst(generatedRequest: HttpRequest): Request[F] = generatedRequest.method match {
  case TrackerMethod.Post(_) =>  Request[F](...)
  ...
}

def address(_) = _
def body(_) = _ 
def parseHeaders(_) = _
```